keps/1714-fair-sharing/README.md

```diff
@@ -0,0 +1,375 @@
+2C should be given an equal share of the resources that correspond to their needs.
+
+#### Story 3
+I have 3 cohorts inside my organization (top cohort): CS, C1 and C2. CS contains
+a single CQ that is given 300 GPU nominal quota. C1 contains 3 CQ: 1A, 1B and 1C,
+among which 1C is considered to be 3 times more important than others. C2 contains
+only one CQ: 2A.  Each of the queues 1A, 1B, 1C and 2A has a long backlog of jobs
+to execute. I want to distribute the resources from CS in the following way:
+
+* Both C1 and C2 get 150 GPUs each, because they are considered equally important.
+* Inside C1, out of these 150 GPUs, 1A gets 30, 1B gets 30 and 1C gets 90 GPUs. 
+1A and 1B are equally important but 1C is 3x more important than 1A and 1B.
+* Inside C2 there is just a single CQ, so it gets all 150 GPUS.
+
+### Risks and Mitigations
+
+* Fair sharing may increase the number of preemptions in the system vs current
+state where the first workload to acquire unused resources keeps them until it
+finishes. Mitigations include:
+    * Introduce minimum execution time before workloads can be preempted for fair sharing.
+    * Introduce delayed fair share enforcement - new workloads have to wait a bit before preempting others to get their share.
+
+* Fair sharing may decrease utilization of the unused resources while attempting to
+distribute them fairly, vs provide the tighties bin-packing. To avoid these scenarios,
+users should prefer to run massive workloads under nominal quotas.
+
+
+## Design Details
+
+The enforcement of fair-share revolves around the idea of a share value per CQ or
+Cohort and mechanisms to keep the share values among siblings in the hierarchy to
+be as equal as possible. This is achieved via two mechanisms:
+
+* During admission, prioritize admitting workload from the ClusterQueues and cohorts that have the lowest share value.
+* Preempt workloads from ClusterQueues and cohorts that have the highest share value.
+
+### Share value function and weights
+
+The value function is a variation of DRF (see
+[1](https://amplab.cs.berkeley.edu/wp-content/uploads/2011/06/Dominant-Resource-Fairness-Fair-Allocation-of-Multiple-Resource-Types.pdf),
+[2](https://dash.harvard.edu/bitstream/handle/1/11956916/Parkes_BeyondDominant.pdf;jsessionid=AC0D06C2CC07C693BD42008D7AE25D99?sequence=1)):
+
+For a given resource r provided by a ClusterQueue or cohort c, we calculate T_r as the total
+requests consumed by the Workloads for that resource in that CQ or cohort, independent of the
+flavor, that are above the nominal quota. The value for a resource is the ratio of T_r and the
+total nominal quotas (or lendingLimits, if defined) in the hierarchy of the parent of C.
+
+The value for the CQ or cohort is the maximum among the values for each resource, divided by the weight, if defined.
+
+Weights will be added to ClusterQueueSpec and CohortSpec in the following optional struct:
+
+```go
```

Comment:
As I can remember, we have a plan to have DRF as another queueingStarategy like this: https://docs.google.com/document/d/1VQ0qxWA-jwgvLq_WYG46OkXWW00O6q7b1BsR_Uv-acs/edit?usp=sharing

So, did you compare the pros and cons of the following options?

1. Exptend CohortSpec and ClusterQueueSpec (current your approach)
2. Introduce a new queueing strategy, "DRF", and have a new CRD, "FairSharing".

@mwielgus @alculquicondor

Reply:
That makes sense.
Recording this discussion as an Alternative approach might be worth it, right?